### ISSUE TITLE
Mechanism to hook in to generated server code with objects generated at runtime for Request Logging

### ISSUE THREAD

**Issue by IvanMalison** [uber/tchannel-go#261]
> Hi @IvanMalison, can you give examples of how you expect this to be used?

See https://code.uberinternal.com/diffusion/EVGOL/browse/master/requestlogger/request_logger.go;44392961db1b3250fb8d2fdb7caceba9add04db4$85

Basically, it should be possible to intercept the arguments that are received by an arbitrary call into the Server object in a uniform way. The existing `Handle` interface does not suffice for this because the only interface it provides is a `thrift.TProtocol` object which can only be read once.

> I'd like to brainstorm to come up with solutions that:

> don't involve use of empty interfaces, at least not in the generated code

I'm pretty skeptical of any dogma against empty interfaces. We are already using `athrift.TStruct` which might as well be empty interfaces since they just have serialization methods. 

> similar to above, not have to do type cases.

I don't think this approach used any type cases. Are you referring to the type assertions that are used once the types are passed back to handleArgs so they can be passed to their respective handlers?

> no perf impact

> Ideally none, but minimal changes to the generated code

This is a strange constraint.

> If there needs more changes to the generated code, then thrift-gen supports templates, so extra code can be generated if the user wants, and the normal generated code should have just the minimal hook.

I'm not sure that templates will suffice in this case, since the generated code needs to have access to objects that are constructed at runtime, elsewhere in the application.
 

--------------------------------------------------
I'm definitely willing to compromise here, but I'm also very skeptical of aggressive performance optimization without benchmarking to back up the idea that performance is a real issue. Do we really think that a signficant portion of a services cpu cycles will be consumed by marshalling between interfaces{} and concrete struct types -- I'd be very surprised to find that the changes that I've made here have even a 10% overall performance impact on even the leanest of services. Premature optimization is the root of all evil and all that stuff. I realize that tchannel-go needs to be quite lean because it is something that will be consumed by absolutely everyone, but for the same reason, it is critically important that the interfaces it provides are convenient and clean. 

An alternative approach that could work to solve the problem of request logging is to add a Method

```AddRequestHook(hook func (ctx Context, args interface{}, resp interface{}, unhandled error))```

to the TChanServer Interface that would populate an instance variable on the server that the generated code would check for and call where appropriate.

The biggest problem with this approach from my perspective is that it makes it much more difficult (possibly impossible) to do something like uncaught exception handling with recover.

**Comment by prashantv** [comment]
Thanks for the use case.

> I'm pretty skeptical of any dogma against empty interfaces. We are already using athrift.TStruct which might as well be empty interfaces since they just have serialization methods.

>I don't think this approach used any type cases. Are you referring to the type assertions that are used once the types are passed back to handleArgs so they can be passed to their respective handlers?

Sorry, I meant "type assertions", not "type case". The combination of empty interfaces and type assertions leads to some ugly code. It's trying to make Go work like a dynamic language, and has the following pitfalls:
 - The empty interface says nothing useful. thrift.TStruct is exactly enough for the other methods, since that's all the library needs is something to serialize.
 - The type assertions can cause panics if misused. The whole point of a compiler is to check these type constraints statically.

>> Ideally none, but minimal changes to the generated code

> This is a strange constraint.

I want to reduce the amount of generated code, and have as much of the logic in the library as possible, unless there's performance or interface reasons not to. Less generated code is easier to maintain for everyone, easier to test and is better handled by tooling. It's also a lot riskier than making library changes since the code in the library is unit tested 100% as is -- that's not true for generated code. It may be possible to test some cases, but depending on the interface, the generated code can change and possibly behave differently. E.g. one example is on the return side, there's void, void with exception, typed return, typed return with exception.

> I'm also very skeptical of aggressive performance optimization without benchmarking to back up the idea that performance is a real issue

I'm not saying this change is going to cause a significant perf issue, but I expect that you should have the data to back up that it isn't a significant perf issue. I'm not going to performance test your code to prove it's slow, I expect you to show that it's it's not slow. 

> I'm not sure that templates will suffice in this case, since the generated code needs to have access to objects that are constructed at runtime, elsewhere in the application.

Not as is, but I'm wondering whether something like Go 1.6's block templates can be used to allow you to add custom hooks in through templates. E.g. if instead there were natural hook in points for pre-read, or post-read inside, that passed parameters like service name, method name etc, then the template could decide whether to use a generic function for all of these cases.

The main issue I have with templates is that middleware then becomes hard to share. People will just copy+paste templates and modify them, instead of layering them. I'm not sure how much of an issue that is. It might be a good starting point though to see how people use it before changing the interfaces.

**Comment by prashantv** [comment]
Discussed offline and came up with an approach to have middleware.

The `thrift.Server` type would allow registering interceptors, which look like the following:

```go
type Interceptor interface {
  // If an error is returned, no further interceptors are called, and the Post method of any
  // previous interceptors is called with the error.
  Pre(ctx Context, method string, args thrift.TStruct) error

  // Post is called when we have a response (or error). If a panic is caught, then the panic is wrapped
  // in thrift/PanicErr. Post may translate response/err into a different error (e.g. replace app.Err with thrift_app.Err).
  Post(ctx Context, method string, arg, response thrift.TStruct, err error) error
}
```

For interceptors that need some state between requests, we could support request-specific interceptors using something like:
```go
type InterceptorFactory interface {
	Create(ctx Context, method string) Interceptor
}
```

One good use case is metrics, but since TChannel already emits metrics for latency, there's no immediate use cases, so we'll defer on `InterceptorFactory` till we really need it.

We should keep the generated code as lightweight as possible, and have most of the logic of running interceptors be in the `thrift` package.

**Comment by IvanMalison** [comment]
@prashantv the meta handler in tchannel-go seems to have some hand coded differences from what gets generated by thrift-gen. This presents a problem because I have added methods to the interface, and changed the format of the code. should I just manually make edits to that file?

**Comment by prashantv** [comment]
@IvanMalison yeah manually changing it probably the easiest option. The `meta` code is not the same as generated code for a couple of reasons:
 - we have to avoid circular dependencies, thrift can't rely on meta, if meta relies on tchannel/thrift
 - we have to put the meta TChannel code inside of tchannel/thrift to avoid above issue, but we don't want to export the meta server, so we manually unexport the identifiers.
